### PR TITLE
RFC: Add convenience methods Shell::info|warn|success()

### PR DESCRIPTION
**Pull request by rchavik**
I had this in my AppShell in 2.x, maybe useful for 3.x.

### CODE REVIEW

**Comment by dereuromark** [comment]
After https://github.com/cakephp/cakephp/commit/072c0ce4cbb1a382e00173e54e20655eb6cbf12e it is sure OK to also provide these wrappers.

PS: Doc blocks missing.

**Comment by rchavik** [comment]
@dereuromark which docblock are you referring to?

**Comment by dereuromark** [comment]
All of them are invalid.

**Comment by markstory** [comment]
This looks good, but it needs tests. Public API methods need tests :smile:

File: src/Console/Shell.php [hunk 1]
```diff
@@ -640,6 +640,48 @@ public function err($message = null, $newlines = 1)
     }
 
     /**
+     * Convenience method for out() that wraps message between <info /> tag
+     *
+     * @param string|array|null $message A string or an array of strings to output
+     * @param int $newlines Number of newlines to append
+     * @param int $level The message's output level, see above.
+     * @return int|bool Returns the number of bytes returned from writing to stdout.
```
Comment (markstory) [inline src/Console/Shell.php]: These return values should be void. Or return the number of bytes.

**Comment by rchavik** [comment]
Now updated with tests :)

**Comment by ADmad** [comment]
Hmm.. Time related tests are failing.

**Comment by rchavik** [comment]
I don't that error is due to this change.  Same exact error is raised in dereuromark's PR with totally different change: https://travis-ci.org/cakephp/cakephp/jobs/98154039

File: src/Console/Shell.php [hunk 2]
```diff
@@ -640,6 +640,48 @@ public function err($message = null, $newlines = 1)
...
+     * @see http://book.cakephp.org/3.0/en/console-and-shells.html#Shell::out
+     */
+    public function warn($message = null, $newlines = 1, $level = Shell::NORMAL)
+    {
+        return $this->out('<warning>' . $message . '</warning>', $newlines, $level);
```
Comment (markstory) [inline src/Console/Shell.php]: Should warnings go to stderr? Or should that be reserved for actual errors?

Comment (dereuromark) [inline src/Console/Shell.php]: I think a warning is often more a user warning type of thing and better be left in stdout.
It simply highlights in a color so you can more easily notice it.

Comment (ionas) [inline src/Console/Shell.php]: @dereuromark Python disagrees with you:
https://docs.python.org/2/library/warnings.html

Comment (ionas) [inline src/Console/Shell.php]: As does NodeJS https://nodejs.org/api/console.html

**Comment by dereuromark** [comment]
I guess that depends on what you do, if it is a technical issue or just user notification of some sorts.

**Comment by ionas** [comment]
Yes, the default should be stderr, configurable to stdout.

**Comment by rchavik** [comment]
Fixed up to use `err()` for `warn()`.

**Comment by markstory** [comment]
Nice work @rchavik

**Comment by dereuromark** [comment]
This is a BC break: https://travis-ci.org/dereuromark/cakephp-cache/jobs/107771130
Are we OK with that, and how can we prevent this in the future?
I mean, most of those wrapper methods should not be exposed to the outside (as commands...).
There I think commands should be suffixed (as do actions):

- infoCommand()
- indexAction()

Especially with moving away from protected for underscore, but also in general this would avoid these BC breaks and API conflicts with internal methods vs expoded public API connection.

**Comment by dereuromark** [comment]
Some other API method collisions present in Console/Shell are
- clear
- abort
- createFile
- helper

etc

For controller actions the currently "known" conflicts are
- components
- redirect
- render
- referer
- paginate
- isAction
- other callbacks

All of the above would be gone if we used suffixing.

**Comment by rchavik** [comment]
That's unfortunate.  But I suppose I am okay with this since we already define/announce it in the migration guide.

**Pull request merged**